TribalWars/src/Infrastructure/Repository/BuildingRequiredRepository.cs

```diff
@@ -0,0 +1,80 @@
+﻿using ApplicationCore.Entities;
+using ApplicationCore.Enums;
+using ApplicationCore.Interfaces;
+using Infrastructure.Data;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infrastructure.Repository
+{
+    public class BuildingRequiredRepository : BaseRepository, IBuildingRequiredRepository
+    {
+        public BuildingRequiredRepository(ApplicationDbContext context) : base(context)
+        {
+        }
+
+        public bool HasMaterial(int buildingId, int level, int villageId)
+        {
+            var requiredMaterial = RequiredMaterials(level, buildingId);
+            var playerMaterials = _context.VillageMaterials.Where(material => material.VillageId == villageId).ToList();
+            if (playerMaterials.Count < requiredMaterial.Count())
+            {
+                return false;
+            }
+            foreach (var material in playerMaterials)
+            {
+                var required = requiredMaterial.FirstOrDefault(m => m.MaterialId == material.MaterialId);
+                if(material.Quantity < required.Quantity)
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public bool HasRequiredBuilding(int buildingId, int villageId)
+        {
+            var requiredBuilding = RequiredBuilding(buildingId);
+            var villageBuilding = _context.VillageBuildings.Where(building => building.VillageId == villageId).ToList();
+            foreach (var building in villageBuilding)
+            {
+                var required = requiredBuilding.FirstOrDefault(b => b.BuildingId == building.BuildingId);
+                if (required == null)
+                {
+                    return false;
+                }
+                else if(required.Level > building.CurrentLevel)
+                {
+                    return false;
+                }
```

Comment:
Same case as with materials. Loop over required buildings, and compare every foundExistingBuilding to every required one